From e4eeab165d0bd4ffc96b3ac9c3bf894247511f0a Mon Sep 17 00:00:00 2001
From: Hardik T Shah <hardik.t.shah@intel.com>
Date: Wed, 9 Mar 2016 18:48:28 +0530
Subject: [PATCH 3835/4301] SoundWire: Add API to register/un-register slave
 driver

commit 9af13930dea575308604a18f0fac5a0cffad9261 from
git://git@github.com:01org/linux-intel-4.9.git

This patch adds API to register/un-register SoundWire slave
device driver.

Change-Id: I76732b273ec31fcd34bd819a1384cbf482a997fc
Signed-off-by: Hardik T Shah <hardik.t.shah@intel.com>
---
 drivers/sdw/sdw.c |   34 ++++++++++++++++++++++++++++++++++
 1 files changed, 34 insertions(+), 0 deletions(-)

diff --git a/drivers/sdw/sdw.c b/drivers/sdw/sdw.c
index 638d273..ec13536 100644
--- a/drivers/sdw/sdw.c
+++ b/drivers/sdw/sdw.c
@@ -1017,6 +1017,40 @@ void sdw_mstr_driver_unregister(struct sdw_mstr_driver *driver)
 }
 EXPORT_SYMBOL_GPL(sdw_mstr_driver_unregister);
 
+void sdw_slave_driver_unregister(struct sdw_slave_driver *driver)
+{
+	driver_unregister(&driver->driver);
+}
+EXPORT_SYMBOL_GPL(sdw_slave_driver_unregister);
+
+/*
+ * An sdw_driver is used with one or more sdw_slave (slave) nodes to access
+ * sdw slave chips, on a bus instance associated with some sdw_master.
+ */
+int __sdw_slave_driver_register(struct module *owner,
+					struct sdw_slave_driver *driver)
+{
+	int res;
+	/* Can't register until after driver model init */
+	if (unlikely(WARN_ON(!sdw_bus_type.p)))
+		return -EAGAIN;
+
+	/* add the driver to the list of sdw drivers in the driver core */
+	driver->driver.owner = owner;
+	driver->driver.bus = &sdw_bus_type;
+
+	/* When registration returns, the driver core
+	 * will have called probe() for all matching-but-unbound slaves.
+	 */
+	res = driver_register(&driver->driver);
+	if (res)
+		return res;
+	pr_debug("sdw-core: driver [%s] registered\n", driver->driver.name);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(__sdw_slave_driver_register);
+
 static void sdw_exit(void)
 {
 	device_unregister(&sdw_slv);
-- 
1.7.5.4

